### PR TITLE
(DNM) Clear TimeField within DatePicker Popover

### PR DESCRIPTION
**Pull request by yihuiliao**
Closes

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

NOTE: Timezones will get cleared if you leave the popover but I believe that's a separate tickets that being worked on [here](https://github.com/adobe/react-spectrum/pull/4715)

Go to any DatePicker story (given that there is a TimeField but you can also change the granularity)
Open up the Popover and scroll to the TimeField
Press backspace twice on either the hour/minute segment
You should get no error and it should look empty

If you leave the popover while the TimeField is empty, the DatePicker should set to its default
If you clear the TimeField and enter a new value, this new value should be set across the entire DatePicker

### CODE REVIEW

File: packages/@react-stately/datepicker/src/useDatePickerState.ts [hunk 1]
```diff
@@ -113,8 +113,14 @@ export function useDatePickerState<T extends DateValue = DateValue>(props: DateP
   };
 
   let selectTime = (newValue: TimeValue) => {
```
Comment (yihuiliao) [inline packages/@react-stately/datepicker/src/useDatePickerState.ts:115]: there are a lot of different ways i could go about writing this conditional so i was wondering if there was a particular way that we'd like to do this?

File: packages/@react-stately/datepicker/src/useDatePickerState.ts [hunk 2]
```diff
@@ -113,8 +113,14 @@ export function useDatePickerState<T extends DateValue = DateValue>(props: DateP
...
-    if (selectedDate && newValue) {
-      commitValue(selectedDate, newValue);
+    if (selectedDate) {
```
Comment (yihuiliao) [inline packages/@react-stately/datepicker/src/useDatePickerState.ts:116]: there are a lot of ways this conditional could be written so i was wondering if there was a particular way we'd like to do this?

**Review by ktabors — APPROVED** [review]

**Comment by snowystinger** [comment]
With the new behavior, I noticed I can get into a state where there is a time in the popover, but not outside of it, and i can't edit the time in the popover.

1. https://reactspectrum.blob.core.windows.net/reactspectrum/bcba9833de7196e69c058999986d1271b0928240/storybook/index.html?path=/story/date-and-time-datepicker--date-time-value&providerSwitcher-express=false
2. delete the timefield segments
3. open the popover/picker
4. type in a time (should see it mirrored in the field i think?)
5. hit esc
6. no time appears outside the popover (probably not expected)
7. reopen the popover/picker, there is a time (not expected)

you can do it with mouse as well, clicking off to close the popover

**Comment by yihuiliao** [comment]
> With the new behavior, I noticed I can get into a state where there is a time in the popover, but not outside of it, and i can't edit the time in the popover.
> 
> 1. https://reactspectrum.blob.core.windows.net/reactspectrum/bcba9833de7196e69c058999986d1271b0928240/storybook/index.html?path=/story/date-and-time-datepicker--date-time-value&providerSwitcher-express=false
> 2. delete the timefield segments
> 3. open the popover/picker
> 4. type in a time (should see it mirrored in the field i think?)
> 5. hit esc
> 6. no time appears outside the popover (probably not expected)
> 7. reopen the popover/picker, there is a time (not expected)
> 
> you can do it with mouse as well, clicking off to close the popover

ohh good catch i'll have to look into that.

i don't think if you're typing in a time within the popover that it should be mirrored in the field? or at least i don't think that's something that's been introduced by this. for example, if you clear the dates so that it's just mm/dd/yyyy and then go and select a date within the popover, you won't see it mirrored in the field until you click outside the popover.

**Review by LFDanLu — COMMENTED** [review]
I haven't looking into it too deeply, but is it possible to preserve the external datefield date when the timefield inside the popover is cleared? Currently it looks like this when you clear the timefield in the popover:
image

Just feels a bit weird to have placeholders for the date when it is still defined.

**Comment by yihuiliao** [comment]
> Just feels a bit weird to have placeholders for the date when it is still defined.

yeah, i think this current solution isn't really gonna work and looking into it a bit more, i think the larger issue is that `selectedTime` never gets set to `null` when you're within the popover (or at least that's how it seems to be)

**Comment by yihuiliao** [comment]
We decided that we will need to do a potential refactor for DatePicker/TimeField which will need to be further discussed after the shutdown. 

There are a couple of oddities that we would like to address to provide a better user experience:
1. If a segment of the datefield is deleted, the value of the date filed becomes null. As a result, anything in the popover is deselected. In instances where the date is filled out but not the time, the entire popover is cleared/unselected. However, we would expect that the date at least be selected and vice versa. 
2. If a popover’s timefield is cleared, the cleared value should persist within the datefield and not be set to the default 12:00 AM
3. If a user fills out a timefield partially (like 10:—), we would like to see that value persist within the popover rather than just null

To accomplish this, we would need to update useDatePickerState to track partial values/segments rather than final values/segments. That might using the logic within useDateFieldState and applying that idea to useDatePickerState.

**Pull request closed without merging**